local/sdk-sriov/pkg/k8s/deviceplugin/server.go

```diff
@@ -0,0 +1,321 @@
+	if err := manager.RegisterDeviceServer(s.ctx, &pluginapi.RegisterRequest{
+		Version:      pluginapi.Version,
+		Endpoint:     socket,
+		ResourceName: s.resourceName,
+	}); err != nil {
+		logEntry.Error("error registering server")
+		return err
+	}
+
+	if err := s.monitorKubeletRestart(manager, socket); err != nil {
+		logEntry.Warnf("error monitoring kubelet restart: %+v", err)
+	}
+
+	return nil
+}
+
+func (s *devicePluginServer) monitorKubeletRestart(manager K8sManager, socket string) error {
+	logEntry := log.Entry(s.ctx).WithField("devicePluginServer", "monitorKubeletRestart")
+
+	resetCh, err := manager.MonitorKubeletRestart(s.ctx)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		logEntry.Infof("start monitoring kubelet restart")
+		defer logEntry.Infof("stop monitoring kubelet restart")
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			case _, ok := <-resetCh:
+				if !ok {
+					return
+				}
+				logEntry.Info("re registering server")
+				if err = manager.RegisterDeviceServer(s.ctx, &pluginapi.RegisterRequest{
+					Version:      pluginapi.Version,
+					Endpoint:     socket,
+					ResourceName: s.resourceName,
+				}); err != nil {
+					logEntry.Errorf("error re registering server: %+v", err)
+					return
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (s *devicePluginServer) GetDevicePluginOptions(_ context.Context, _ *pluginapi.Empty) (*pluginapi.DevicePluginOptions, error) {
+	return &pluginapi.DevicePluginOptions{}, nil
+}
+
+func (s *devicePluginServer) ListAndWatch(_ *pluginapi.Empty, server pluginapi.DevicePlugin_ListAndWatchServer) error {
+	logEntry := log.Entry(s.ctx).WithField("devicePluginServer", "ListAndWatch")
+
+	for {
+		resp, err := s.resourceListerClient.List(s.ctx, &podresources.ListPodResourcesRequest{})
+		if err != nil {
+			logEntry.Errorf("resourceListerClient unavailable: %+v", err)
+			return err
+		}
+
+		s.updateDevices(s.respToDeviceIds(resp))
+
+		if err := server.Send(s.listAndWatchResponse()); err != nil {
+			logEntry.Errorf("server unavailable: %+v", err)
+			return err
+		}
+
+		select {
+		case <-s.ctx.Done():
+			logEntry.Info("server stopped")
+			return s.ctx.Err()
+		case <-time.After(s.resourcePollTimeout):
+		case <-s.updateDevicesCh:
+		}
+	}
+}
+
+func (s *devicePluginServer) respToDeviceIds(resp *podresources.ListPodResourcesResponse) map[string]bool {
+	deviceIds := map[string]bool{}
+	for _, pod := range resp.PodResources {
+		for _, container := range pod.Containers {
+			for _, device := range container.Devices {
+				if device.ResourceName == s.resourceName {
+					for _, id := range device.DeviceIds {
+						deviceIds[id] = true
+					}
+				}
+			}
+		}
+	}
+	return deviceIds
+}
+
+func (s *devicePluginServer) updateDevices(idsInUse map[string]bool) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	var freeCount int
+	for id, device := range s.devices {
+		switch {
+		case idsInUse[id]:
+			device.free = false
+		case !device.free:
+			device.free = true
```

Comment:
Why we mark the device as free here and do not call cleanup?

Reply:
We could have made a mistake marking this device as free. So on the first update we mark it as free, on the second we clean it up if it is still free.

Reply:
Why we can make a mistake marking this device as free?

Reply:
There possibly can be such case:

dps.ListAndWatch(...) -> []{ "1", "2", "3", "4" }
kubelet.List(...) -> []{ "1", "2", "3" }
**dps.Allocate([]{ "3", "4" }) -> OK, got it, no problem, you can use "4" device**
\- Ok, we should try to cleanup devices, what did we get from kubelet?
\- []{ "1", "2", "3" }
\- Nice, let's mark "4" device as free and delete it!
dps.ListAndWatch(...) -> []{ "1", "2", "3" } **but there is a container using "4" device, so it would be stopped**

The point is to mark "4" device as free but avoid deleting it before we double check that it is actually free so we really can do it.